From e5412178c132a5f66eaf2e2fa1f88a8f5e0e54b5 Mon Sep 17 00:00:00 2001
From: Cyril Chemparathy <cyril@ti.com>
Date: Mon, 10 Dec 2012 13:24:04 -0500
Subject: [PATCH 01/10] uio: add support for ioctls

- This patch introduces an ioctl handler in uio_info,
allowing for basic ioctl plumbing to underlying drivers.

Signed-off-by: Cyril Chemparathy <cyril@ti.com>

- This is needed for allowing uio based drivers to support ioctl calls.
This will allow customized calls from user space applications to
the uio based drivers.

- Changed default return value from -ENOTSUPP to -ENOTTY

Signed-off-by: Sam Nelson <sam.nelson@ti.com>
---
 drivers/uio/uio.c          | 13 +++++++++++++
 include/linux/uio_driver.h |  2 ++
 2 files changed, 15 insertions(+)

diff --git a/drivers/uio/uio.c b/drivers/uio/uio.c
index 65bf067..46e9fde 100644
--- a/drivers/uio/uio.c
+++ b/drivers/uio/uio.c
@@ -704,6 +704,18 @@ static int uio_mmap(struct file *filep, struct vm_area_struct *vma)
 	}
 }
 
+static long uio_ioctl(struct file *filep, unsigned int cmd, unsigned long arg)
+{
+	struct uio_listener *listener = filep->private_data;
+	struct uio_device *idev = listener->dev;
+	long ret = -ENOTTY;
+
+	if (idev->info->ioctl)
+		ret = idev->info->ioctl(idev->info, cmd, arg);
+
+	return ret;
+}
+
 static const struct file_operations uio_fops = {
 	.owner		= THIS_MODULE,
 	.open		= uio_open,
@@ -714,6 +726,7 @@ static const struct file_operations uio_fops = {
 	.poll		= uio_poll,
 	.fasync		= uio_fasync,
 	.llseek		= noop_llseek,
+	.unlocked_ioctl	= uio_ioctl,
 };
 
 static int uio_major_init(void)
diff --git a/include/linux/uio_driver.h b/include/linux/uio_driver.h
index 32c0e83..8ead26b 100644
--- a/include/linux/uio_driver.h
+++ b/include/linux/uio_driver.h
@@ -90,6 +90,7 @@ struct uio_device {
  * @open:		open operation for this uio device
  * @release:		release operation for this uio device
  * @irqcontrol:		disable/enable irqs when 0/1 is written to /dev/uioX
+ * @ioctl:		ioctl operation for this uio device
  */
 struct uio_info {
 	struct uio_device	*uio_dev;
@@ -105,6 +106,7 @@ struct uio_info {
 	int (*open)(struct uio_info *info, struct inode *inode);
 	int (*release)(struct uio_info *info, struct inode *inode);
 	int (*irqcontrol)(struct uio_info *info, s32 irq_on);
+	long (*ioctl)(struct uio_info *info, unsigned cmd, unsigned long arg);
 };
 
 extern int __must_check
-- 
1.9.1

